model/metrics/vcs.yaml

```diff
@@ -0,0 +1,150 @@
+groups:
+  - id: metric.vcs.repository.change.count
+    type: metric
+    metric_name: vcs.repository.change.count
+    brief: 'The number of changes (pull requests) in a repository, categorized by their state (either open or merged)'
+    instrument: gauge
+    unit: "{change}"
+    stability: experimental
+    #extends: metric_attributes.vcs
+    attributes:
+      - ref: vcs.change.state
+        requirement_level: required
+      - ref: vcs.repository.name
+        requirement_level: required
+  - id: metric.vcs.repository.change.time_open
+    type: metric
+    metric_name: vcs.repository.change.time_open
+    brief: 'The amount of time a change (pull request) has been open'
+    instrument: gauge
+    unit: "s"
+    stability: experimental
+    attributes:
+      - ref: vcs.repository.name
+        requirement_level: required
+      - ref: vcs.repository.ref.name
+        requirement_level: required
+  - id: metric.vcs.repository.change.time_to_approval
+    type: metric
+    metric_name: vcs.repository.change.time_to_approval
+    brief: 'The amount of time it took a change (pull request) to go from open to approved'
+    instrument: gauge
+    unit: "s"
+    stability: experimental
+    attributes:
+      - ref: vcs.repository.name
+        requirement_level: required
+      - ref: vcs.repository.ref.name
+        requirement_level: required
+  - id: metric.vcs.repository.change.time_to_merge
+    type: metric
+    metric_name: vcs.repository.change.time_to_merge
+    brief: 'The amount of time it took a change (pull request) to go from open to merged'
+    instrument: gauge
+    unit: "s"
+    stability: experimental
+    attributes:
+      - ref: vcs.repository.name
+        requirement_level: required
+      - ref: vcs.repository.ref.name
+        requirement_level: required
+  - id: metric.vcs.repository.count
```

Comment:
is it really necessary as a metric? I assume you'd report it daily and it'd be wiser/more frugal to report it as an event or just call GH API whenever you need this info

Reply:
These metrics in general all come from a component that scrapes the GH API. The API returns a list of repositories to iterate on because to gather the other metrics through the API, one has to know the repository.  I don't see why we wouldn't include it as a valid metric, but I would say it's not a required metric. Thoughts?

Reply:
Is this the component we own and created? 
There are multiple ways to write it - e.g. using webhooks vs scraping.

All of the GH things are really rare - even huge monorepos get maybe 100s of PRs opened a day. 

If you have a hook a report every even as it arrives (even per-comment!) it's maybe 1000s of events per day. If you exclude comments/issues (that are not yet in the scope  of this PR), it'd probably be in 100s order of magnitude.
So it's cheap.

Reporting them as events allows you to include high-cardinality data, do any aggregation that you want at the query time.  E.g. you might want to report contributor (or it's anonymized name) as an attribute, or number of comments the PR received - it'd allow to derive almost anything.

You'd need an event for just a few things - repo, PR, issue, comment, branch lifecycles - maybe ~10 distinct events.
------------------------

Metrics, if you scrape them periodically (e.g. once an hour), would report 24 points for each metric a day and there are 12  metrics already (not even including issues or comments).  It's 300 points a day.

But you already pre-aggregated everything and some information got lost - you have to change the component code to create yet another metric that users would be able to report if they wanted to. 

----------------------

TL;DR:
- if VCS things were reported as events
  - it would still be cheap (low volume)
  - very flexible (you aggregate them at query time)
  - could have been reported in real time from hooks (low volume)
  - you would have easy-to-understand things like "pr_opened", "pr_closed" 
  - you could write collector processor to aggregate them to metrics so that dashboards scale better. The cool part users write any aggregation they want or not aggregate at all.
- if VCS things are reported as metrics
  - it's at least as expensive as events (low volume, but metrics are reported periodically), 
  - rigid (you need to envision every aggregation users want to do), 
  - you'd end up with a lot of very specific metrics like repo count or number of lines of code, it's a bigger mental load to understand all that's available
  - reported rarely (API scrapes are more expensive then hooks)  

So I strongly suggest to revisit the approach and consider using events instead of metrics. 

Also it seems GH would suggest using webhooks here https://docs.github.com/en/webhooks/about-webhooks#choosing-webhooks-or-the-rest-api


If it's some 3rd party component that emits metrics and we can't change them - we had something similar in the past around legacy Kafka metrics which were not quite aligned with OTel - we ended up removing them from semantic conventions and recommended documenting them in the corresponding collector receiver:

- https://github.com/open-telemetry/semantic-conventions/issues/269
- https://github.com/open-telemetry/semantic-conventions/pull/196
- https://github.com/open-telemetry/semantic-conventions/pull/113

Reply:
Events are not a catch all be all. There are valid reasons to use scraping for **some** of the metrics here. There are certainly metrics I've found to be effective that comes from the GitHub event system.

I wrote this component & contributed it to OTEL. I am happy to make adjustments (traces from events are in the works, and I would also like to add transformed events from my WebHook Event receiver + transform processor solution in the future). 

To the original question, I think it'd be ok to change some to histogram. I also don't think these are "required metrics." The purpose of my comment was to bring context & seek to understand. The metrics originate out of openo11y.dev & dora capabilities.

Reply:
We decided that currently defined metrics provide a good basis.
In the future we might want to expand further and use events.
But to be discussed on another issue / PR.